domain/resourcegroup/runaway.go

```diff
@@ -75,84 +79,291 @@ type RunawayRecord struct {
 		EndTime:           endTime,
 		Watch:             watchType,
 		WatchText:         convict,
 		Source:            rm.serverID,
-	}:
+		Action:            action,
+	}
+	rm.addWatchList(record, ttl, false)
+	select {
+	case rm.quarantineChan <- record:
 	default:
 		// TODO: add warning for discard flush records
 	}
 }
+
+func (rm *RunawayManager) addWatchList(record *QuarantineRecord, ttl time.Duration, force bool) {
+	key := record.GetRecordKey()
+	// This is a pre-check, because we generally believe that in most cases, we will not add a watch list to a key repeatedly.
+	item := rm.getWatchFromWatchList(key)
+	if force {
+		rm.queryLock.Lock()
+		defer rm.queryLock.Unlock()
+		if item != nil {
+			if item.ID == record.ID {
+				return
+			}
+			rm.watchList.Delete(key)
+		}
+		rm.watchList.Set(key, record, ttl)
+	} else {
+		if item == nil {
+			rm.queryLock.Lock()
+			// When watchlist get record, it will check whether the record is stale, so add new record if returns nil.
+			if rm.watchList.Get(key) == nil {
+				rm.watchList.Set(key, record, ttl)
+			} else {
+				rm.staleQuarantineRecord <- record
+			}
+			rm.queryLock.Unlock()
+		} else if item.ID != record.ID {
+			rm.staleQuarantineRecord <- record
+		}
+	}
+}
+
+// GetWatchByKey is used to get a watch item by given key.
+func (rm *RunawayManager) GetWatchByKey(key string) *QuarantineRecord {
+	return rm.getWatchFromWatchList(key)
+}
+
+// GetWatchList is used to get all watch items.
+func (rm *RunawayManager) GetWatchList() []*QuarantineRecord {
+	items := rm.watchList.Items()
+	ret := make([]*QuarantineRecord, 0, len(items))
+	for _, item := range items {
+		ret = append(ret, item.Value())
+	}
+	return ret
+}
+
+// AddWatch is used to add watch items from system table.
+func (rm *RunawayManager) AddWatch(record *QuarantineRecord) {
+	ttl := time.Until(record.EndTime)
+	if record.EndTime.Equal(NullTime) {
+		ttl = 0
+	} else if ttl <= 0 {
+		rm.staleQuarantineRecord <- record
+		return
+	}
+
+	force := false
+	// The manual record replaces the old record.
+	force = record.Source == ManualSource || record.Source == rm.serverID
+	rm.addWatchList(record, ttl, force)
+}
+
+// RemoveWatch is used to remove watch items from system table.
+func (rm *RunawayManager) RemoveWatch(record *QuarantineRecord) {
+	// we should check whether the cached record is not the same as the removing record.
+	rm.queryLock.Lock()
+	defer rm.queryLock.Unlock()
+	item := rm.getWatchFromWatchList(record.GetRecordKey())
+	if item == nil {
+		return
+	}
+	if item.ID == record.ID {
+		rm.watchList.Delete(record.GetRecordKey())
+	}
+}
+func (rm *RunawayManager) getWatchFromWatchList(key string) *QuarantineRecord {
+	item := rm.watchList.Get(key)
+	if item != nil {
+		return item.Value()
+	}
+	return nil
+}
+
 func (rm *RunawayManager) markRunaway(resourceGroupName, originalSQL, planDigest string, action string, matchType RunawayMatchType, now *time.Time) {
+	source := rm.serverID
+	if len(source) > 128 {
```

Comment:
Why need truncate here?

Reply:
I think the server IP in k8s will be domain name, so the length of it will be very long. And the length of feild is 128, do we need to change it to blob?

Reply:
You should set a field length that is long enough to store the data

Reply:
OK, let's set 512 same as analyze job